enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/snapshot/PersistentSnapshotDownloader.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.causalclustering.core.state.snapshot;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.LockSupport;
+
+import org.neo4j.causalclustering.catchup.storecopy.StoreCopyFailedException;
+import org.neo4j.causalclustering.core.consensus.LeaderLocator;
+import org.neo4j.causalclustering.core.consensus.NoLeaderFoundException;
+import org.neo4j.causalclustering.core.state.CommandApplicationProcess;
+import org.neo4j.causalclustering.helper.ExponentialBackoffStrategy;
+import org.neo4j.causalclustering.helper.TimeoutStrategy;
+import org.neo4j.logging.Log;
+
+class PersistentSnapshotDownloader implements Runnable
+{
+    private final CommandApplicationProcess applicationProcess;
+    private final LeaderLocator leaderLocator;
+    private final CoreStateDownloader downloader;
+    private final Log log;
+    private final TimeoutStrategy.Timeout timeout;
+    private State state;
+
+    PersistentSnapshotDownloader( LeaderLocator leaderLocator,
+            CommandApplicationProcess applicationProcess, CoreStateDownloader downloader, Log log,
+            TimeoutStrategy.Timeout timeout )
+    {
+        this.applicationProcess = applicationProcess;
+        this.leaderLocator = leaderLocator;
+        this.downloader = downloader;
+        this.log = log;
+        this.timeout = timeout;
+        this.state = State.INITIATED;
+    }
+
+    PersistentSnapshotDownloader( LeaderLocator leaderLocator,
+            CommandApplicationProcess applicationProcess, CoreStateDownloader downloader, Log log )
+    {
+        this( leaderLocator, applicationProcess, downloader, log,
+                new ExponentialBackoffStrategy( 1, 30, TimeUnit.SECONDS ).newTimeout() );
+    }
+
+    private enum State
+    {
+        INITIATED,
+        RUNNING,
+        COMPLETED
+    }
+
+    @Override
+    public void run()
+    {
+        state = State.RUNNING;
+        try
+        {
+            applicationProcess.pauseApplier( CoreStateDownloaderService.OPERATION_NAME );
```

Comment:
This will not work correctly on subsequent attempts after failure because you will pause several times, but only resume once on the last successful attempt. The pause/resume functionality is reference counting so you will be left with a reference count > 0. You need an object with a lifetime of all the attempts to pause and resume, once each.